File: GDPIControl/FormBlacklist.vb
Imports System.Net
Imports System.IO

Public Class FormBlacklist
	Private WithEvents Client As New WebClient()

	Private Async Sub FormBlacklist_Load(sender As Object, e As EventArgs) Handles MyBase.Load
		Try
			If File.Exists(Consts.NewBlacklistPath) Then File.Delete(Consts.NewBlacklistPath)
			Await Client.DownloadFileTaskAsync(New Uri(Consts.BlacklistURI), Consts.NewBlacklistPath)
			If File.Exists(Consts.BlacklistPath) Then File.Delete(Consts.BlacklistPath)
			File.Move(Consts.NewBlacklistPath, Consts.BlacklistPath)
			MessageBox.Show("Blacklist download done")
		Catch ex As Exception
			MessageBox.Show(ex.Message)
		End Try
	End Sub

	Private Sub Client_DownloadProgressChanged(sender As Object, e As DownloadProgressChangedEventArgs) Handles Client.DownloadProgressChanged
		PB.Value = e.ProgressPercentage
		L_Size.Text = $"{Math.Round(e.TotalBytesToReceive / 1024, 2)} КВ"
		L_Done.Text = $"{Math.Round(e.BytesReceived / 1024, 2)} КВ"
	End Sub

	Private Sub FormBlacklist_FormClosed(sender As Object, e As FormClosedEventArgs) Handles Me.FormClosed

	End Sub

End Class

File: GDPIControl/FormBlacklist.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Windows.Forms;

namespace GDPIControl
{
    public partial class FormBlacklist : Form
    {
        private readonly WebClient Client = new();
        private readonly Dictionary<RadioButton, string> RBtoURI = new();

        public FormBlacklist()
        {
            InitializeComponent();
            RBtoURI.Add(RB_AZ, @"https://antizapret.prostovpn.org/domains-export.txt");
            RBtoURI.Add(RB_AZM, @"https://mirror.virenbar.workers.dev/?key=blacklist-gdpi");
            RBtoURI.Add(RB_AF, @"https://antifilter.download/list/domains.lst");

            Client.Headers.Add("Origin", "GDPIControl");
            Client.DownloadProgressChanged += Client_DownloadProgressChanged;
        }

        private void Client_DownloadProgressChanged(object sender, DownloadProgressChangedEventArgs e)
        {
            PB.Value = e.ProgressPercentage;
            L_Size.Text = $"{Math.Round((double)e.TotalBytesToReceive / 1024, 2)} КВ";
            L_Done.Text = $"{Math.Round((double)e.BytesReceived / 1024, 2)} КВ";
        }

        #region UI Events

        private async void B_Download_Click(object sender, EventArgs e)
        {
            B_Download.Enabled = false;
            try
            {
                var uri = RBtoURI.First(KV => KV.Key.Checked).Value;
                if (File.Exists(Constants.BlacklistTempPath)) { File.Delete(Constants.BlacklistTempPath); }
                await Client.DownloadFileTaskAsync(new Uri(uri), Constants.BlacklistTempPath);
                if (File.Exists(Constants.BlacklistPath)) { File.Delete(Constants.BlacklistPath); }
                File.Move(Constants.BlacklistTempPath, Constants.BlacklistPath);
                this.ShowInfo("Blacklist download done");
            }
            catch (Exception E)
            {
                this.ShowError(E, "Download Error");
            }
            finally
            {
                B_Download.Enabled = true;
            }
        }

        private void FormBlacklist_FormClosed(object sender, FormClosedEventArgs e)
        {
            if (Client.IsBusy) { Client.CancelAsync(); }
            Client.Dispose();
        }

        #endregion UI Events
    }
}